DQMServices/Core/interface/locking_ptr.h

```diff
@@ -0,0 +1,136 @@
+#ifndef locking_ptr_h
+#define locking_ptr_h
+
+#include <memory>
+
+namespace detail {
+
+  class lock_base
+  {
+  public:
+    lock_base() = default;
+    virtual ~lock_base() = default;
+
+    virtual void lock() const = 0;
+    virtual void unlock() const = 0;
+  };
+
+  template <typename L>
+  class lock_impl : public lock_base
+  {
+  public:
+    lock_impl<L>(L & lock) :
+      lock_(lock)
+    { }
+
+    void lock() const final
+    {
+      lock_.lock();
+    }
+
+    void unlock() const final
+    {
+      lock_.unlock();
+    }
+
+  private:
+    L & lock_;
+  };
+
+
+  template <typename T>
+  class locking_ptr_impl
+  {
+  public:
+    locking_ptr_impl<T>(T * object, lock_base * lock) :
+      object_(object),
+      lock_(lock)
+    {
+      lock_->lock();
+    }
+
+    ~locking_ptr_impl<T>()
+    {
+      lock_->unlock();
+    }
+
+    T * operator->() {
+      return object_;
+    }
+
+    const T * operator->() const {
+      return object_;
+    }
+
+  protected:
+    T * object_;
+    lock_base * lock_;
+  };
+
+} // detail
+
+template <typename T>
+class locking_ptr
+{
+public:
+  locking_ptr<T>() :
+    object_(nullptr),
```

Comment:
This is needed to allow creating a vector of this type, and othet usage patterns I found in users' code.